### PR TITLE
Parameter layer

### PR DESCRIPTION
**Pull request by daniyalaliev**
Layer with trainable parameters, which goes to output while inference (nn.Parameter analogue)

### CODE REVIEW

File: NeoML/test/src/ParameterLayerTest.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/* Copyright � 2017-2020 ABBYY Production LLC
```
Comment (favorart) [inline NeoML/test/src/ParameterLayerTest.cpp]: New format of 1st line copyright is
/* Copyright © < Year-created >-< Year-updated > ABBYY
So for new files
/* Copyright © 2024 ABBYY

All files should be saved in UTF8 codepage.

File: NeoML/src/Dnn/Layers/ParameterLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+namespace NeoML {
+
+void CParameterLayer::AllocateOutputBlobs()
+{
+	outputBlobs[0] = paramBlobs[0];
+}
+
+void CParameterLayer::SetBlob(CDnnBlob* _blob)
+{
+	if (_blob == paramBlobs[0].Ptr()) {
```
Comment (favorart) [inline NeoML/src/Dnn/Layers/ParameterLayer.cpp]: Maybe, I'm not sure, the check should also be here, that blob.ReinterpretDimensions() on the changing blob is not be called outside. And the blob still has the same dimensions.

File: NeoML/src/Dnn/Layers/ParameterLayer.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	if (!outputBlobs.IsEmpty()) {
+		outputBlobs[0] = 0;
+	}
+}
+
+void CParameterLayer::Reshape()
+{
+	CheckOutputs();
```
Comment (favorart) [inline NeoML/src/Dnn/Layers/ParameterLayer.cpp:69]: Should it be a check inputsCount == 0?

**Review by favorart — APPROVED** [review]

File: NeoML/include/NeoML/Dnn/Layers/ParameterLayer.h [hunk 4]
```diff
@@ -29,13 +29,16 @@ class NEOML_API CParameterLayer : public CBaseLayer {
...
+    void SetBlobDesc(const CBlobDesc& _desc);
     const CPtr<CDnnBlob>& GetBlob() const { return paramBlobs[0]; }
 protected:
     void AllocateOutputBlobs() override;
     void Reshape() override;
     void RunOnce() override;
     void BackwardOnce() override;
     void LearnOnce() override;
+private:
+    CBlobDesc desc;
```
Comment (favorart) [inline NeoML/include/NeoML/Dnn/Layers/ParameterLayer.h:41]: Need to add `desc` field to `serialize` method

File: NeoML/src/Dnn/Layers/ParameterLayer.cpp [hunk 5]
```diff
@@ -47,12 +47,33 @@ void CParameterLayer::SetBlob(CDnnBlob* _blob)
...
+void CParameterLayer::SetBlobDesc(const CBlobDesc& _desc)
+{
+	bool isReshapeNeeded = desc.GetDataType() == CT_Invalid
+		|| !desc.HasEqualDimensions(_desc)
+		|| desc.GetDataType() != _desc.GetDataType();
+
+	desc = _desc;
+
+	if( isReshapeNeeded ) {
+		ForceReshape();
```
Comment (favorart) [inline NeoML/src/Dnn/Layers/ParameterLayer.cpp:60]: Need to drop 'paramBlobs[0] = 0;' to reinitialize it before `ForceReshape` call

**Review by SAngeliuk — APPROVED** [review]

**Pull request merged**